examples/02_model_collaborative_filtering/cornac_bpr_deep_dive.ipynb

```diff
@@ -88,16 +88,38 @@
     "\n",
     "### 1.1 Personalized Ranking from Implicit Feedback\n",
     "\n",
-    "The task of personalized ranking aims at providing each user a ranked list of items (recommendations).  This is very common in scenarios where recommender systems are based on implicit user behavior (e.g. purchases, clicks).  The available observations are only positive feedback where the non-observed ones are a mixture of real negative feedback and missing values.\n",
+    "The task of personalized ranking aims at providing each user a ranked list of items (recommendations). This is very common in scenarios where recommender systems are based on implicit user behavior (e.g. purchases, clicks). The available observations are only positive feedback where the non-observed ones are a mixture of real negative feedback and missing values.\n",
     "\n",
-    "One usual approach for item recommendation is directly predicting a preference score $\\hat{x}_{u,i}$ given to item $i$ by user $u$.  BPR uses a different approach by using item pairs $(i, j)$ and optimizing for the correct ranking given preference of user $u$, thus, there are notions of *positive* and *negative* items.  The training data $D_S : U \\times I \\times I$ is defined as:\n",
+    "One usual approach for item recommendation is directly predicting a preference score $\\hat{x}_{u,i}$ given to item $i$ by user $u$. This method uses the **pointwise loss** to optimize the model. Methods based on pointwise learning usually follow a regression framework by minimizing the  squared loss between $\\hat{x}_{u,i}$ its target value ${x}_{u,i}$. Unfortunately, this method doesn't work well with ranking.\n",
+    "\n",
+    "<p align=\"center\">\n",
+    "<img src=\"https://recodatasets.z20.web.core.windows.net/images/bpr1.png\" width=\"400px\">\n",
+    "</p>\n",
+    "\n",
+    "In the implicit feedback scenario the observed data is positive feedback. As it is shown in the figure above, the observed data (image to the left) is known, but the unobserved data is unknown. The negative data is usually generated by filling the matrix with 0 (image to the right). \n",
+    "\n",
+    "Using the pointwise loss, the model is trained to predict as 1 the elements in the observed dataset and 0 for the rest. The problem with this approach is that all elements the model should rank in the future are presented as negative feedback during training. Therefore, an overfitted model, a model that learns the training data exactly, will predict all new elements as 0s. The only reason these models can predict rankings is by using techniques to prevent overfitting, like regularization.\n",
```

Comment:
This doesn't hold in general: "an overfitted model, a model that learns the training data exactly, will predict all new elements as 0s". Some predictions may be zeros, other ones, depending on the algorithm and the hyperparameters. Also, filling the rest of the matrix with zeros is not what happens in a collaborative filtering scenario, since there is nothing left to predict.

Reply:
This comes from the paper: "Machine learning approaches for item recommenders
[5, 10] typically create the training data from S by
giving pairs (u, i) ∈ S a positive class label and all
other combinations in (U × I) \ S a negative one (see
Figure 1). Then a model is fitted to this data. That
means the model is optimized to predict the value 1 for
elements in S and 0 for the rest. The problem with this
approach is that all elements the model should rank in
the future ((U × I) \ S) are presented to the learning
algorithm as negative feedback during training. That
means a model with enough expressiveness (that can
fit the training data exactly) cannot rank at all as it
predicts only 0s. The only reason why such machine
learning methods can predict rankings are strategies
to prevent overfitting, like regularization."
https://arxiv.org/ftp/arxiv/papers/1205/1205.2618.pdf

do you think this doesn't hold true? A perfect network will predict everything unknown as zero, because we are saying that both negatives and unknowns are zero. Regularization is what makes the trick so this doesn't happen in practice.

Reply:
Yes, I think what they are saying is that some methods like the SVD method of [13] will overfit (because you specify the whole matrix and interpolate on that). But other methods like [5] (which is similar to ALS) use regularization, so the reg. parameters can be selected so as to avoid overfitting.
I am not sure this says much though. After all the BPR method will overfit too (in a different way) if you set the lambda_Theta to zero. The main difference from methods like [5] is the loss function. So, with BPR you can overfit the pairs of items i.e. overfit the differences in sec. 4.3 (which, up to an additive constant, implies that you overfit the observed ones and zeros).

Reply:
I guess what I am saying is that overfitting is not an issue just with pointwise loss, it will also appear with BPR if you set the reg. param to zero. The text in the notebook gives another impression though.

Reply:
got it, makes sense. I have modified the text based on what you said. Let me know if it is more accurate now